docs/path-resolution.rst

```diff
@@ -0,0 +1,690 @@
+.. _path-resolution:
+
+***************
+Path Resolution
+***************
+
+In order to be able to support reproducible builds on all platforms, the Solidity compiler has to
+abstract away the details of the filesystem where source files are stored.
+Paths used in imports must work the same way everywhere while the command-line interface must be
+able to work with platform-specific paths to provide good user experience.
+This section aims to explain in detail how Solidity reconciles these requirements.
+
+.. index:: ! virtual filesystem, ! source unit name
+.. _virtual-filesystem:
+
+Virtual Filesystem
+==================
+
+The paths used in :ref:`imports <import>` in a general case do not have to be filesystem paths.
+The compiler maintains an internal database (virtual filesystem) where each compiled source unit is
+assigned a unique *source unit name* which is an opaque and unstructured identifier.
+
+Source files can be placed in the virtual filesystem directly, using :ref:`Standard JSON interface
+<compiler-api>` and in this case source unit names can be anything.
+In most cases, however, the project files reside in the local filesystem and it is desirable for
+the compiler to be able to locate and load them automatically.
+To support this, the compiler passes all source unit names not found in the virtual filesystem to
+the file loader.
+In case of the command-line compiler the file loader attempts to interpret them as local paths.
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ is a bit more flexible in that
+regard and allows the user to provide a callback, which can interpret the source unit name in an
+arbitrary way.
+For example load it from the network if it is an URL.
```

Comment:
Maybe: "Remix uses this to load files from github."

Reply:
I originally did not want to repeat it (it's already mentioned in the section about URLs). But after adding the section on the callback I moved that mention here.